add randomness kwarg to jacfwd

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84220

From https://github.com/pytorch/functorch/issues/1010, if a user runs jacfwd with a function that uses randomness, it will fail since the default behavior for vmap is error. This lets the user specify the randomness behavior to jacfwd too since it is doing vmap(jvp(forward)). This is less likely to show up in jacrev since that only vmaps over the backwards pass

Commit:
add randomness kwarg to jacfwd

[ghstack-poisoned]

Commit:
Update on "add randomness kwarg to jacfwd"


From https://github.com/pytorch/functorch/issues/1010, if a user runs jacfwd with a function that uses randomness, it will fail since the default behavior for vmap is error. This lets the user specify the randomness behavior to jacfwd too since it is doing vmap(jvp(forward)). This is less likely to show up in jacrev since that only vmaps over the backwards pass

[ghstack-poisoned]

Commit:
Update on "add randomness kwarg to jacfwd"


From https://github.com/pytorch/functorch/issues/1010, if a user runs jacfwd with a function that uses randomness, it will fail since the default behavior for vmap is error. This lets the user specify the randomness behavior to jacfwd too since it is doing vmap(jvp(forward)). This is less likely to show up in jacrev since that only vmaps over the backwards pass

[ghstack-poisoned]

zou3519 commented:
What operators do randomness in their backward pass? I would have thought that was a bug

zou3519 commented on functorch/functorch/_src/eager_transforms.py:
```diff
@@ -839,7 +839,7 @@ def safe_unflatten(tensor, dim, shape):
 
 
-def jacfwd(func: Callable, argnums: argnums_t = 0, has_aux: bool = False):
+def jacfwd(func: Callable, argnums: argnums_t = 0, has_aux: bool = False, randomness: str = "error"):
```
Let's start making arguments kwarg only. Does `*, randomness: str = "error"` work?

samdow replied:
Fixed and tests work

zou3519 commented on functorch/functorch/_src/eager_transforms.py:
```diff
@@ -855,6 +855,9 @@ def jacfwd(func: Callable, argnums: argnums_t = 0, has_aux: bool = False):
             auxiliary objects that will not be differentiated.
             Default: False.
+        randomness(str): Flag indicating what type of randomness to use.
+            See :func:`vmap`. Allowed: "different", "same", "error".
```
nit: see :func:`vmap` for more details

Could you check that the docs for this build before merging it? Since we don't have a docs build set up in pytorch/pytorch yet

samdow replied:
Edited and checked docs. Link goes to vmap correctly

<img width="717" alt="Screen Shot 2022-09-06 at 4 24 25 PM" src="https://user-images.githubusercontent.com/17888388/188733544-078f712a-b18a-49f8-8f5f-8c774ed8a2fb.png">

Commit:
Update on "add randomness kwarg to jacfwd"


From https://github.com/pytorch/functorch/issues/1010, if a user runs jacfwd with a function that uses randomness, it will fail since the default behavior for vmap is error. This lets the user specify the randomness behavior to jacfwd too since it is doing vmap(jvp(forward)). This is less likely to show up in jacrev since that only vmaps over the backwards pass

[ghstack-poisoned]

samdow commented:
@pytorchbot merge

samdow commented:
@pytorchbot rebase

samdow commented:
@pytorchbot merge

Commit:
Update on "add randomness kwarg to jacfwd"


From https://github.com/pytorch/functorch/issues/1010, if a user runs jacfwd with a function that uses randomness, it will fail since the default behavior for vmap is error. This lets the user specify the randomness behavior to jacfwd too since it is doing vmap(jvp(forward)). This is less likely to show up in jacrev since that only vmaps over the backwards pass

[ghstack-poisoned]
